micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateQueryMetrics.java

```diff
@@ -0,0 +1,175 @@
+        monitor(registry, sessionFactory, sessionFactoryName, Tags.of(tags));
+    }
+
+    /**
+     * Create {@code HibernateQueryMetrics} and bind to the specified meter registry.
+     *
+     * @param registry           meter registry to use
+     * @param sessionFactory     session factory to use
+     * @param sessionFactoryName session factory name as a tag value
+     * @param tags               additional tags
+     */
+    public static void monitor(MeterRegistry registry, SessionFactory sessionFactory, String sessionFactoryName, Iterable<Tag> tags) {
+        new HibernateQueryMetrics(sessionFactory, sessionFactoryName, tags).bindTo(registry);
+    }
+
+    /**
+     * Create a {@code HibernateQueryMetrics}.
+     *
+     * @param sessionFactory     session factory to use
+     * @param sessionFactoryName session factory name as a tag value
+     * @param tags               additional tags
+     */
+    public HibernateQueryMetrics(SessionFactory sessionFactory, String sessionFactoryName, Iterable<Tag> tags) {
+        this.tags = Tags.concat(tags, SESSION_FACTORY_TAG_NAME, sessionFactoryName);
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+
+        registerListeners(registry);
+
+    }
+
+    /**
+     * Register hibernate loadEvent listener - event trigger the metrics registration,
+     * only select queries are recorded in QueryStatistics {@link org.hibernate.stat.spi.StatisticsImplementor}
+     *
+     * @param meterRegistry meterRegistry to use
+     */
+    private void registerListeners(MeterRegistry meterRegistry) {
+        if (sessionFactory instanceof SessionFactoryImplementor) {
+            EventListenerRegistry registry = ((SessionFactoryImplementor) sessionFactory).getServiceRegistry().getService(EventListenerRegistry.class);
+            MetricsEventHandler metricsEventHandler = new MetricsEventHandler(meterRegistry);
+            registry.appendListeners(EventType.POST_LOAD, metricsEventHandler);
+        }
+
+    }
+
+    class MetricsEventHandler implements PostLoadEventListener {
+
+        private final MeterRegistry meterRegistry;
+
+        MetricsEventHandler(MeterRegistry meterRegistry) {
+            this.meterRegistry = meterRegistry;
+        }
+
+        @Override
+        public void onPostLoad(PostLoadEvent event) {
+            registerQueryMetric(event.getSession().getFactory().getStatistics());
+        }
+
+        void registerQueryMetric(Statistics statistics) {
+            for (String query : statistics.getQueries()) {
+                QueryStatistics queryStatistics = statistics.getQueryStatistics(query);
+                if (Objects.nonNull(queryStatistics)) {
+                    String queryName = meterRegistry.config().namingConvention().tagValue(query);
+                    if (Objects.isNull(Search.in(meterRegistry).tags("query", queryName).functionCounter())) {
```

Comment:
Is the point of this check because there could be multiple of the same query returned by `statistics.getQueries()`?

Reply:
Emm If I remember correctly there should not, it is a little bit redundant check